go/vt/topo/test/shard.go

```diff
@@ -94,3 +96,72 @@ func checkShard(t *testing.T, ts *topo.Server) {
 		t.Errorf("GetShardNames(666): %v", err)
 	}
 }
+
+// checkShardWithLock verifies the Shard operations work correctly With Lock
+func checkShardWithLock(t *testing.T, ts *topo.Server) {
+	ctx := context.Background()
+	if err := ts.CreateKeyspace(ctx, "test_keyspace", &topodatapb.Keyspace{}); err != nil {
+		t.Fatalf("CreateKeyspace: %v", err)
+	}
+
+	unblock := make(chan struct{})
+	finished := make(chan struct{})
+
+	// Check GetShardNames returns [], nil for existing keyspace with no shards.
+	if names, err := ts.GetShardNames(ctx, "test_keyspace"); err != nil || len(names) != 0 {
+		t.Errorf("GetShardNames(keyspace with no shards) didn't return [] nil: %v %v", names, err)
+	}
+
+	if err := ts.CreateShard(ctx, "test_keyspace", "b0-c0"); err != nil {
+		t.Fatalf("CreateShard: %v", err)
+	}
+
+	_, unlock1, err := ts.LockShard(ctx, "test_keyspace", "b0-c0", "lock")
+	if err != nil {
+		t.Errorf("CreateShard called second time, got: %v", err)
+	}
+
+	// As soon as we're unblocked, we try to lock the keyspace.
+	go func() {
+		<-unblock
+		waitUntil := time.Now().Add(10 * time.Second)
+		var isUnLocked1 = false
+		for time.Now().Before(waitUntil) {
+			_, unlock2, err := ts.TryLockShard(ctx, "test_keyspace", "b0-c0", "lock")
+			if err != nil {
+				if !topo.IsErrType(err, topo.NodeExists) {
+					require.Fail(t, "expected node exists during trylock", err.Error())
+				}
+				time.Sleep(1 * time.Second)
+				var finalErr error
+				unlock1(&finalErr)
+				isUnLocked1 = true
+				if finalErr != nil {
+					require.Fail(t, "Unlock(test_keyspace) failed", finalErr.Error())
+				}
+			} else {
+				unlock2(&err)
+				if err != nil {
+					require.Fail(t, "Unlock(test_keyspace) failed", err.Error())
+				}
+				if isUnLocked1 {
+					close(finished)
+				}
+				break
+			}
+		}
+	}()
+
+	// sleep for a while so we're sure the go routine is blocking
+	time.Sleep(2 * time.Second)
```

Comment:
A comment I often get in my own PRs, the magic number used here can cause tests to be flaky. Is it possible to wait _until_ the goroutine is certain to be blocking? Wait for some channel/event?

Reply:
+1. I don't see the point of this sleep. The goroutine will block on writing a message to the non-buffered channel so it will block until we close that channel below. Why sleep? What are we waiting for? Can we check for that?